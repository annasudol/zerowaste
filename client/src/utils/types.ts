export interface LocationTypes {
    hash?: string;
    pathname: string;
    search?: string;
    state: {
        ingredients: string[];
    };
    key?: string;
}
