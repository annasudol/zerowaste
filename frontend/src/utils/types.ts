export interface LocationTypes {
    hash?: string
    pathname: string
    search?: string
    state: {
        ingredients?: string[]
        backPath?: string
        callRefetch?: boolean
    };
    key?: string
}



export interface ListType {
    title: string
    id: number
}


export interface RecipeStateProps {
    title: string
    servings: number
    image: string
    readyInMinutes: number
    ingredients: string[] | []
    detailedIngredients: string[] | []
    instructions: string
    sourceUrl: undefined | string
}