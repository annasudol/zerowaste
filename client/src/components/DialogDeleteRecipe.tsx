import * as React from "react";
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogTitle from '@material-ui/core/DialogTitle';
import Slide from '@material-ui/core/Slide';
import { TransitionProps } from '@material-ui/core/transitions';
import { useMutation, useSubscription } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import { ErrorMessage } from "../components";
import { Button } from 'antd';

const DELETE_RECIPE = gql`
  mutation DeleteRecipe($id: ID!) {
    deleteRecipe(id: $id) {
        id
    }
  }
`;

const DELETE_RECIPE_SUBSCRIPTION = gql`
 subscription deleteRecipe {
  deleteRecipe {
    user {
      name
      email
      recipes{
        title
        id
        image
        ingredients
      }
    }
  }
 }
`;

// tslint:disable-next-line: no-shadowed-variable
const Transition = React.forwardRef(function Transition(
    props: TransitionProps & { children?: React.ReactElement<unknown, string> },
    ref: React.Ref<unknown>,
) {
    return <Slide direction="up" ref={ref} {...props} />;
});

interface DialogDeleteRecipeProps {
    open: boolean
    toggleOpen(val: boolean): void
    recipeId: string
    title: string
}
export const DialogDeleteRecipe: React.FC<DialogDeleteRecipeProps> = ({ open, toggleOpen, recipeId, title }): React.ReactElement => {
    const [deleteRecipe, { error }] = useMutation(DELETE_RECIPE);
    useSubscription(DELETE_RECIPE_SUBSCRIPTION);

    const handleDeleteRecipe = () => {
        deleteRecipe({ variables: { id: recipeId } });
        toggleOpen(false);
    }

    if (error) {
        return (<ErrorMessage message={error.message} />)
    }
    const alert = `Are you sure do you want to delete ${title} recipe`

    return (
        <Dialog
            open={open}
            TransitionComponent={Transition}
            keepMounted
            onClose={(): void => toggleOpen(false)}
            aria-labelledby="alert-dialog-slide-title"
            aria-describedby="alert-dialog-slide-description"
        >
            <DialogTitle id="alert-dialog-slide-title">{alert}</DialogTitle>
            <DialogActions>
                <Button onClick={(): void => toggleOpen(false)}>
                    Keep the recipe
                </Button>

                <Button onClick={handleDeleteRecipe} type="default" danger>
                    Delete recipe
                </Button>
            </DialogActions>
        </Dialog>
    );
}