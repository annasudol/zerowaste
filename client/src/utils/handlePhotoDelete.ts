export const handlePhotoDelete = async (url: string) => {
    const publicId = url.match(/(?=).{20}?(?=.jpe?g|.png)/);
    if (publicId) {
        try {
            const response = await fetch('http://localhost:9000/delete/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ public_id: publicId[0] })
            });
            return response.json();

        } catch { }
    }

}