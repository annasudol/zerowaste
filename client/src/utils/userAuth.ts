const IS_LOGGED_IN = 'isLoggedIn'

export const getAccessToken = () => localStorage.getItem(IS_LOGGED_IN)

export const logoutUser = () => localStorage.removeItem(IS_LOGGED_IN)
