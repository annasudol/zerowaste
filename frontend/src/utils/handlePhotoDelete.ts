export const handlePhotoDelete = async (url: string) => {
    const publicId = url.match(/(?=).{20}?(?=.jpe?g|.png)/);
    if (publicId) {
        try {
            const response = await fetch(`http://localhost:${process.env.REACT_APP_PORT}/delete/`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ public_id: publicId[0] })
            });
            console.log(response, "response")
            return response.json();

        } catch { }
    }

}