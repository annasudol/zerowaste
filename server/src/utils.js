
module.exports.paginateResults = ({
  page = 0,
  pageSize = 10,
  results,
}) => {
  const resultsIntoPages = page * pageSize;
  return results.slice(resultsIntoPages, resultsIntoPages + pageSize);
};
