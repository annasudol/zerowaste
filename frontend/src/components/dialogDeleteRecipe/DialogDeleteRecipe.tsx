import React, { FC, ReactElement } from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogTitle from '@material-ui/core/DialogTitle';
import { useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import { Button } from 'antd';
import { handlePhotoDelete } from '../../utils/handlePhotoDelete';

export const DELETE_RECIPE = gql`
  mutation DeleteRecipe($id: ID!) {
    deleteRecipe(id: $id) {
        id
    }
  }
`;




interface DialogDeleteRecipeProps {
    open: boolean
    toggleOpen(val: boolean): void
    recipeId: string
    title: string
    image: string
}

export const DialogDeleteRecipe: FC <DialogDeleteRecipeProps> = ({open, toggleOpen, recipeId, title, image}): ReactElement => {
    const [deleteRecipe] = useMutation(DELETE_RECIPE, {
        onError(err) {
          console.log(err);
        },
    });

    const handleDeleteRecipe = () => {
        handlePhotoDelete(image)
        deleteRecipe({variables: {id: recipeId } });
        toggleOpen(false);
    }


    return (
        <Dialog
            open={open}
            keepMounted
            aria-labelledby="alert-dialog-slide-title"
            aria-describedby="alert-dialog-slide-description"
        >
        <DialogTitle id="alert-dialog-slide-title">{`Are you sure do you want to delete ${title} recipe`}</DialogTitle>
        <DialogActions>
            <Button onClick={(): void => toggleOpen(false)}>
                Keep the recipe
            </Button>
            <Button onClick={handleDeleteRecipe} type="default" danger>
                Delete recipe
            </Button>
        </DialogActions>
    </Dialog>
    );
}