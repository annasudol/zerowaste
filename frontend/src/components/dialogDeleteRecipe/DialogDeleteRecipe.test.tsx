import React from 'react';
import { DialogDeleteRecipe, DELETE_RECIPE } from './DialogDeleteRecipe';
import * as apollo from '@apollo/react-hooks';
import { renderApollo, cleanup,  fireEvent, waitForElement } from '../../test.utils';
import { act } from "react-dom/test-utils";


describe('component', () => {
  describe('DialogDeleteRecipe', () => {
    afterEach(cleanup);
      it('match snapshots', (): void => {
        const spy = jest.spyOn(apollo, 'useSubscription');

        spy.mockReturnValue({
            variables: { id: "5ef248ca324f6aaa02bfd7a0" },
            loading: false,
        });

        const mocks = [
          {
            request: {
              query: DELETE_RECIPE,
              variables: { id: "5ef248ca324f6aaa02bfd7a0" },
            },
            results: {}
          }
        ]
     const {container} = renderApollo(<DialogDeleteRecipe open toggleOpen={jest.fn()} recipeId="1" title="dialog title" image="myImage" />, {mocks})

     expect(container.firstChild).toMatchSnapshot();
      });

      it('responds for keeping recipe', async () => {
        const spy = jest.spyOn(apollo, 'useSubscription');

        spy.mockReturnValue({
            variables: { id: "5ef248ca324f6aaa02bfd7a0" },
            loading: false,
        });

        const mocks = [
          {
            request: {
              query: DELETE_RECIPE,
              variables: { id: "5ef248ca324f6aaa02bfd7a0" },
            },
            results: {data: { id: "5ef248ca324f6aaa02bfd7a0" }}

          }
        ]
        const mockToggleOpen = jest.fn();
        const {findByText} = renderApollo(<DialogDeleteRecipe open toggleOpen={mockToggleOpen} recipeId="1" title="dialog title" image="myImage" />, {mocks})
        const deleteButton =  await waitForElement(() => findByText(/keep the recipe/i))
        act(()=> {
          fireEvent.click(deleteButton);
        })
        expect(mockToggleOpen).toHaveBeenCalledWith(false);
      });
      it('responds for closing dialog and delete recipe', async () => {

        const spy = jest.spyOn(apollo, 'useSubscription');

        spy.mockReturnValue({
            variables: { id: "5ef248ca324f6aaa02bfd7a0" },
            loading: false,
        });

        const mocks = [
          {
            request: {
              query: DELETE_RECIPE,
              variables: { id: "5ef248ca324f6aaa02bfd7a0" },
              results: {data: { id: "5ef248ca324f6aaa02bfd7a0" }}
            }
          }
        ]
        const mockToggleOpen = jest.fn();
        const {findByText} = renderApollo(<DialogDeleteRecipe open toggleOpen={mockToggleOpen} recipeId="1" title="dialog title" image="myImage" />, {mocks})
        const deleteButton =  await waitForElement(() => findByText(/delete recipe/i))
        act(()=> {
          fireEvent.click(deleteButton);
        })
        expect(mockToggleOpen).toHaveBeenCalledWith(false);
      });
    });
});

