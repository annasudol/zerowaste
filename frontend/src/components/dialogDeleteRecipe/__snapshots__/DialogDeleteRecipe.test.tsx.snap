// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component DialogDeleteRecipe match snapshots 1`] = `null`;
